mautic/mautic
# Improve performance of the frequency rule query

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This improves the frequency rule query by breaking up into two queries, one for those with custom rules and one without. For the default rules, it allows to do simple calculations in PHP (such as the date) rather than asking MySQL to calculate on the fly. 

Original (table with 133 million email stats): 32.8 seconds
```
SELECT ch.lead_id, count(ch.lead_id) 
FROM mautic_email_stats ch 
WHERE (ch.date_sent >= '2018-11-08 21:23:54') AND (ch.lead_id IN (...)) 
GROUP BY ch.lead_id, fr.frequency_time, fr.frequency_number 
HAVING (count(ch.lead_id) >= fr.frequency_number AND ch.lead_id IN (...))
```

Broken into two queries: total 528.7ms
with custom rules: 49.7 ms
```
SELECT ch.lead_id, fr.frequency_number, fr.frequency_time 
FROM mautic_email_stats ch 
INNER JOIN mautic_lead_frequencyrules fr ON ch.lead_id = fr.lead_id 
WHERE fr.channel = 'email' AND ((ch.date_sent >= case fr.frequency_time
                 when 'MONTH' then DATE_SUB(NOW(),INTERVAL 1 MONTH)
                 when 'DAY' then DATE_SUB(NOW(),INTERVAL 1 DAY)
                 when 'WEEK' then DATE_SUB(NOW(),INTERVAL 1 WEEK)
                end)) AND (ch.lead_id IN (...)) 
GROUP BY ch.lead_id, fr.frequency_time, fr.frequency_number 
HAVING count(ch.lead_id) >= fr.frequency_number;
```

default for those without custom rules: 479ms
```
SELECT ch.lead_id
FROM mautic_email_stats ch 
WHERE (ch.date_sent >= '2018-11-08 21:23:54') AND (ch.lead_id IN (...)) 
    AND (NOT EXISTS (SELECT null FROM mautic_lead_frequencyrules fr WHERE fr.lead_id = ch.lead_id)) 
GROUP BY ch.lead_id 
HAVING count(ch.lead_id) >= 3;
``` 

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Define a default email frequency rule in Mautic's configuration on the email settings tab
2. Go to the preference center for a single contact and define a different custom frequency rule (allow more than the default)
3. Add the contact above along with some others into a segment
4. Send emails to these contacts until they hit the frequency rule for both and emails are starting to queue
5. The contact with a custom frequency rule should not have had emails queued until it hit it's limit
6. All other contacts should have started queuing emails once it hit the limit as defined in Mautic's configuration
7. Repeat the test without a global configuration option set and a new contact with custom frequency rules and newly added to the segment)
8. Send emails again and this time only the contact with a custom frequency rule should have mail queued. The rest should be sent since there is no default.
